### ISSUE TITLE
Gen utilities are streams wrapping single values

### ISSUE THREAD

**Issue by disordered** [zio/zio#9146]
Most `Gen` convenience methods are creating generators that are not behaving like streams. In certain situations this causes tests to be flaky and/or consume extra CPU (and time). The reason for this seems to be that the utilities wrap constants and rely on repeated `Gen` resampling, instead of "unfolding" the stream.

To illustrate this, I have created a bit extreme example:
```scala
test("gen filter") {
  val filteredGen = for {
    _ <- Gen.const(true)
    _ = println("Regenerating")
    first <- Gen.int(1, 10).filter(_ == 10)
    second <- Gen.int(1, 10).filter(_ == 1)
    third <- Gen.int(1, 10).filter(_ == 5)
  } yield s"$first.$second.$third"

  assertZIO(filteredGen.runCollectN(1).map(_.headOption))(isSome(equalTo("10.1.5")))
} @@ TestAspect.repeats(10)
```

There are a couple of issues with the above:
1. If assert uses `runHead` instead of `runCollectN`, it will almost certainly fail the test
2. Every filtered generator compounds the probability of it returning a `None`

If above would be coded using `ZStream.range`, calling `runHead` on it, would never fail to produce a value and my expectation is that generators should act the same way.
As for compounding failure rate, the test would succeed if it would be used inside a `check`, because behind the scenes, it uses a forever repeating sampling. However, this costs a lot of repeat regenerations, as it will only yield a value when all the combined generators happen to produce a non-filtered value. In other words, if at least one of them produces a value that is filtered out, the generator will run from the beginning.

The above test prints `Regenerating` on average ~1000+ times, instead of once.
If superfluous `Gen.const(true)` and `println` are removed, it takes about 300ms to complete on my machine.

However, if I slightly modify the `filteredGen` to the following:
```scala
val filteredGen = for {
  _ <- Gen.const(true)
  _ = println("Regenerating")
  first <- Gen(Gen.int(1, 10).sample.forever).filter(_ == 10)
  second <- Gen(Gen.int(1, 10).sample.forever).filter(_ == 1)
  third <- Gen(Gen.int(1, 10).sample.forever).filter(_ == 5)
} yield s"$first.$second.$third"
```

The `runHead` will always produce a value.
`Regenerating` will be printed only once.
Test completes on average in about 100ms on my machine (excluding the `Gen.const` and `println`).

Specifically for `Gen.int`, this is because it is created as a `ZStream.fromZIO`. Alphanumeric generators suffer from the same. I didn't check all of them, but I suspect the issue is present in a lot of the functions.

I see two solutions to this:
1. Change to instead use `ZStream.unfold`
2. Change to wrap `sample.forever`

I haven't looked into the codebase well enough to know what effects this would have, so I will hold off on submitting a PR. However, if some guidance can be provided and the fix is not overly involving, I could take it up.

**Comment by guizmaii** [comment]
Have you tried with this approach:
```scala
test("gen filter") {
  val filteredGen = for {
    _ <- Gen.const(true)
    _ = println("Regenerating")
    first <- Gen.int(1, 10).filter(_ == 10)
    second <- Gen.int(1, 10).filter(_ == 1)
    third <- Gen.int(1, 10).filter(_ == 5)
  } yield s"$first.$second.$third"

  assertZIO(filteredGen.sample.forever.take(1))(equalTo("10.1.5"))
} @@ TestAspect.repeats(10)
```

I did a similar change here: https://github.com/zio/zio-json/pull/1152

**Comment by disordered** [comment]
> Have you tried with this approach:
> 
> ```scala
> test("gen filter") {
>   val filteredGen = for {
>     _ <- Gen.const(true)
>     _ = println("Regenerating")
>     first <- Gen.int(1, 10).filter(_ == 10)
>     second <- Gen.int(1, 10).filter(_ == 1)
>     third <- Gen.int(1, 10).filter(_ == 5)
>   } yield s"$first.$second.$third"
> 
>   assertZIO(filteredGen.sample.forever.take(1))(equalTo("10.1.5"))
> } @@ TestAspect.repeats(10)
> ```
> 
> I did a similar change here: [zio/zio-json#1152](https://github.com/zio/zio-json/pull/1152)

It is the same as using `runCollectN`: https://github.com/zio/zio/blob/76818b4f80eacb601f45029ba0abd93c33a9fe77/test/shared/src/main/scala/zio/test/Gen.scala#L156-L157

As I mentioned above, it works in a sense that test won't fail, but it suffers from performance hit when you combine multiple filtered generators. It also doesn't address `runHead` not working. It is what I am using right now in my project.

**Comment by guizmaii** [comment]
I'll be honest. Either I don't understand your issue or I'm not convinced it's not a Gen issue but your usage of them which is incorrect.

**Comment by jdegoes** [comment]
This seems to be part of the larger problem that sampling and finite generators are really different entities and there exists points in ZIO Test where one semantic or the other is desired.

**Comment by joroKr21** [comment]
That is more or less on purpose, see #9101 - if we use streams with > 1 element then combining multiple generators would behave like `ZStream.flatMap` which means not really random at all. The best practice is to use correctly narrow generators instead of using `filter` - e.g. `Gen.const(10)` instead of `Gen.int(1, 10).filter(_ == 10)` etc.

**Comment by disordered** [comment]
As John said, I am probably misusing the ZIO test generators. I'm not using them for property based testing - I thought this would be an easy way to generate unique, complex objects/values for integration testing. The example with filtering to constant `Int` was just for illustration of `runHead` and performance issues that I experience.
In addition, ZIO test runs all tests in the suite on the same random seed (instead of reporting seed which failed the test), which forces me to wrap all these `Gen` calls into `live(...)`. So, it seems like I should find a different approach and this issue can be rejected.

**Comment by joroKr21** [comment]
> The example with filtering to constant Int was just for illustration of runHead and performance issues that I experience.

I think the performance issues you are experiencing come from using `filter`, not from generators in general. You can imagine that it takes more work to find a valid sample after generating candidates instead of just generating valid candidates all the time. If you can avoid using `filter`, you shouldn't really notice it.

> In addition, ZIO test runs all tests in the suite on the same random seed (instead of reporting seed which failed the test)

You can use `TestAspect.nondeterministic` to vary the seed. I'm not sure about reporting it.

**Comment by disordered** [comment]
> I think the performance issues you are experiencing come from using `filter`, not from generators in general. You can imagine that it takes more work to find a valid sample after generating candidates instead of just generating valid candidates all the time. 

Yes, it is due to `filter` usage, but my initial point was that current implementation makes combined generators with filters significantly slower, since failure rate compounds with each filter usage in the combined generator. Besides, it also breaks `runHead`.

**Comment by joroKr21** [comment]
@disordered what if I write it like this?
Should it be regenerating `first` or not?
```scala
      test("gen filter doesn't backtrack") {
        val filteredGen = for {
          first  <- Gen.int(1, 10)
          second <- Gen.int(1, 10).filter(_ == 1)
          if first == 10
          third  <- Gen.int(1, 10).filter(_ == 5)
        } yield s"$first.$second.$third"
        assertZIO(filteredGen.runHead)(isSome(equalTo("10.1.5")))
      } @@ TestAspect.repeats(10)
```

**Comment by joroKr21** [comment]
I don't think both can work, but maybe it's better to not regenerate because you can move the filter around.
Anyway, I propose a solution in #9378

**Comment by disordered** [comment]
> @disordered what if I write it like this? Should it be regenerating `first` or not?
> 
> ```scala
>       test("gen filter doesn't backtrack") {
>         val filteredGen = for {
>           first  <- Gen.int(1, 10)
>           second <- Gen.int(1, 10).filter(_ == 1)
>           if first == 10
>           third  <- Gen.int(1, 10).filter(_ == 5)
>         } yield s"$first.$second.$third"
>         assertZIO(filteredGen.runHead)(isSome(equalTo("10.1.5")))
>       } @@ TestAspect.repeats(10)
> ```

No, it should not, since the guard is on the second generator. The above example is equivalent to
```scala
Gen.int(1, 10)
  .flatMap { first => 
    Gen.int(1, 10)
      .filter(_ == 1)
      .withFilter(_ => first == 10)
      .flatMap { second =>
        Gen.int(1, 10)
          .filter(_ == 5)
          .map(third => s"$first.$second.$third")
      }
  }
```
If you would move the guard to first generator, then it should take next sample from `first`.

In other words, `Gen` is a specialized stream (unless I'm assuming wrong here), so I would expected it to behave like one.
